src/Components/Shared/src/JsonSerialization/JsonConverterFactoryTypeInfoResolver.cs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Text.Json.Serialization.Metadata;
+
+namespace Microsoft.AspNetCore.Components;
+
+// For custom converters that don't rely on serializing an object graph,
+// we can resolve the incoming type's JsonTypeInfo directly from the converter.
+// This skips extra work to collect metadata for the type that won't be used.
+internal sealed class JsonConverterFactoryTypeInfoResolver<T> : IJsonTypeInfoResolver
```

Comment:
Depending on how many `JsonConverterFactoryTypeInfoResolver<T>` instances you want to create in your application, you might want to consider bundling them altogether into a non-generic `JsonConverterFactoryTypeInfoResolver` that simply accepts an allow list `Type[]` you want handled by this resolver.

Reply:
That's actually what I had initially, but reduced it down to this generic type for two reasons:
1. So far, we've only had to do this for `DotNetObjectReference`.
2. By having access to the generic type, we can directly invoke `JsonMetadataServices.CreateValueInfo<T>()` without using reflection.

We could probably work around limitation 2 by starting with an empty set of supported types and adding to it via some `JsonConverterFactoryTypeInfoResolver.AddType<T>()` method which would then map `typeof(T)` to its respective `JsonMetadataServices.CreateValueInfo<>()` delegate. But at that point I'm not sure that's particularly better than having multiple `JsonConverterFactoryTypeInfoResolver<T>`s in a single chain.

Reply:
Even if irestricted to just one type, the class still doesn't need to be generic and could just accept a single `Type` parameter via the constructor. I'm approaching this from the perspective that generic types add more overhead to the runtime(s), so they should be removed unless necessary.